sw/device/lib/base/print.h

```diff
@@ -0,0 +1,94 @@
+ * format.
+ */
+typedef void (*buffer_sink_t)(const char *, size_t);
+
+/**
+ * Prints out a message to stdout, formatted according to the format string
+ * |format|.
+ *
+ * The definition of "stdout" is not provided by this library; rather, it must
+ * be initialized using |base_set_stdout()|.
+ *
+ * This function supports a subset of the format specifiers provided by standard
+ * C |printf|. Those are, namely:
+ * - %%, which prints a percent sign.
+ * - %c, which prints a single ASCII character.
+ * - %s, which prints a NUL-terminated string.
+ * - %d and %i, which print a signed decimal uint32_t.
+ * - %u, which prints an unsigned decimal uint32_t.
+ * - %o, which prints an unsigned octal uint32_t.
+ * - %x and %X, which print an unsigned hex uint32_t.
+ * - %n, which prints the number of bytes printed so far.
+ * - %p, which prints a pointer in an unspecified way.
+ *
+ * Additionally, three SystemVerilog format specifiers are supported:
+ * - %h and %H, which are aliases for %x and %X, respectively.
+ * - %b, which prints an unsigned binary uint32_t.
+ *
+ * Finally, an additional nonstandard format specifier is supported:
+ * - %z, which takes a size followed by a buffer, and prints out |size|
+ *   characters from the buffer.
+ *
+ * This function furthermore supports width modifiers for integer specifiers,
+ * such as |%10d|. It does not support dynamic widths like |%*d|, and will also
+ * always pad with zeroes, rather than spaces.
+ *
+ * Of course, providing arguments for formatting which are incompatible with a
+ * given format specifier is Undefined Behavior.
+ *
+ * @arg format the format spec.
+ * @arg ... values to interpolate in the format spec.
+ */
+size_t base_printf(const char *format, ...);
+
+/**
+ * Prints a message to the sink |out|.
+ *
+ * See |base_printf()| for the semantics of the format specification.
+ *
+ * @arg out a sink to print to.
+ * @arg format the format spec.
+ * @arg ... values to interpolate in the format spec.
+ */
+size_t base_fprintf(buffer_sink_t out, const char *format, ...);
+
+/**
+ * Prints a message to the sink |out|.
+ *
+ * This function is identical to |base_fprintf|, except in that it takes a
+ * |va_list| instead of having a vararg parameter. This function is provided
+ * not for calling directly, but rather for being called by functions that
+ * already take a variable number of arguments, and wish to make use of
+ * formatting facilities.
+ *
+ * This function takes ownership of |args|.
+ *
+ * @arg out a sink to print to.
+ * @arg format the format spec.
+ * @arg args values to interpolate in the format spec.
+ */
+size_t base_core_printf(buffer_sink_t out, const char *format, va_list args);
+
+/**
+ * Sets what the "stdout" sink is, which is used by |base_printf()|.
+ *
+ * The default stdout sink simply throws away its arguments, so
+ * |base_set_stdout()| must be called before printed messaged wind up anywhere.
```

Comment:
"…, so |base_set_stdout()| must be called if the user wants the printed messages to go anywhere."